models/buffer_disk.go

```diff
@@ -0,0 +1,196 @@
+package models
+
+import (
+	"fmt"
+	"os"
+	"sync"
+
+	"github.com/influxdata/telegraf"
+	"github.com/influxdata/telegraf/metric"
+	"github.com/tidwall/wal"
+)
+
+type DiskBuffer struct {
+	BufferStats
+	sync.Mutex
+
+	walFile     *wal.Log
+	walFilePath string
+
+	batchFirst uint64 // index of the first metric in the batch
+	batchSize  uint64 // number of metrics currently in the batch
+}
+
+func NewDiskBuffer(name string, path string, stats BufferStats) (*DiskBuffer, error) {
+	filePath := path + "/" + name
+	walFile, err := wal.Open(filePath, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to open wal file: %w", err)
+	}
+	return &DiskBuffer{
+		BufferStats: stats,
+		walFile:     walFile,
+		walFilePath: filePath,
+	}, nil
+}
+
+func (b *DiskBuffer) Len() int {
+	b.Lock()
+	defer b.Unlock()
+	return b.length()
+}
+
+func (b *DiskBuffer) length() int {
+	// Special case for when the read index is zero, it must be empty (otherwise it would be >= 1)
+	if b.readIndex() == 0 {
+		return 0
+	}
+	return int(b.writeIndex() - b.readIndex())
+}
+
+// readIndex is the first index to start reading metrics from, or the head of the buffer
+func (b *DiskBuffer) readIndex() uint64 {
+	index, err := b.walFile.FirstIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index
+}
+
+// writeIndex is the first index to start writing metrics to, or the tail of the buffer
+func (b *DiskBuffer) writeIndex() uint64 {
+	index, err := b.walFile.LastIndex()
+	if err != nil {
+		panic(err) // can only occur with a corrupt wal file
+	}
+	return index + 1
+}
+
+func (b *DiskBuffer) Add(metrics ...telegraf.Metric) int {
+	b.Lock()
+	defer b.Unlock()
+
+	dropped := 0
+	for _, m := range metrics {
+		if !b.addSingle(m) {
+			dropped++
+		}
+	}
+	b.BufferSize.Set(int64(b.length()))
+	return dropped
+	// todo implement batched writes
+}
+
+func (b *DiskBuffer) addSingle(m telegraf.Metric) bool {
+	data, err := m.ToBytes()
+	if err != nil {
+		panic(err)
+	}
+	err = b.walFile.Write(b.writeIndex(), data)
+	m.Accept()
+	if err == nil {
+		b.metricAdded()
+		return true
+	}
+	return false
+}
+
+//nolint:unused // to be implemented in the future
+func (b *DiskBuffer) addBatch(metrics []telegraf.Metric) int {
+	written := 0
+	batch := new(wal.Batch)
+	for _, m := range metrics {
+		data, err := m.ToBytes()
+		if err != nil {
+			panic(err)
+		}
+		m.Accept() // accept here, since the metric object is no longer retained from here
+		batch.Write(b.writeIndex(), data)
+		b.metricAdded()
+		written++
+	}
+	err := b.walFile.WriteBatch(batch)
+	if err != nil {
+		return 0 // todo error handle, test if a partial write occur
+	}
+	return written
```

Comment:
We cannot do this! Either we panic here if we decide this is the right thing to do or we do proper error handling. 

Imagine a use-case where it is crucial to collect _all_ metrics and not loose any and your disk runs full. The user might have a kafka instance for buffering. Now you accepted the metrics in the for-loop above but then writing fails here due to the disk being full... There will be no error, Telegraf will show zero metrics written (just as if no metric arrived at the output) but the metrics are removed from Kafka because we accepted them. Now imaging the fun debugging this situation! :-)